### PR TITLE
Introducing isComplete method to receivers

### PR DESCRIPTION
**Pull request by supunkamburugamuve**
Changing the semantics of op receivers to

progress() - return true if further progres is needed
isComplete() - return true if the receiver is synced

### CODE REVIEW

File: twister2/api/src/java/edu/iu/dsc/tws/api/comms/BaseOperation.java [hunk 1]
```diff
@@ -105,4 +93,13 @@ public boolean progressChannel() {
   public boolean sendBarrier(int src, byte[] barrierId) {
     return this.op.send(src, barrierId, MessageFlags.SYNC_BARRIER);
   }
+
+  /**
+   * Wait for the operation to complete
+   */
+  public void waitForCompletion() {
+    while (op.isComplete()) {
```
Comment (chathurawidanage) [inline twister2/api/src/java/edu/iu/dsc/tws/api/comms/BaseOperation.java]: while(op.isComplete()) or while(!op.isComplete())

Comment (supunkamburugamuve) [reply twister2/api/src/java/edu/iu/dsc/tws/api/comms/BaseOperation.java]: Thanks, this is fixed

File: twister2/comms/src/java/edu/iu/dsc/tws/comms/utils/OperationUtils.java [hunk 2]
```diff
@@ -31,6 +31,41 @@
...
+      if (partialLock.tryLock()) {
+        try {
+          mergeComplete = partialReceiver.isComplete();
+        } finally {
+          partialLock.unlock();
+        }
+      }
+    } catch (Throwable t) {
+      LOG.log(Level.SEVERE, "un-expected error", t);
+      throw new RuntimeException(t);
```
Comment (chathurawidanage) [inline twister2/comms/src/java/edu/iu/dsc/tws/comms/utils/OperationUtils.java]: Twister2RuntimeException()

File: twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/partition/DPartitionBatchFinalReceiver.java [hunk 3]
```diff
@@ -296,26 +301,27 @@ public boolean progress() {
         lock.unlock();
       }
     }
-    return finishedTargetsCompleted.size() != targets.size();
+    boolean b = finishedTargetsCompleted.size() != targets.size();
+    complete = !b;
+    return b;
```
Comment (chathurawidanage) [inline twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/partition/DPartitionBatchFinalReceiver.java]: can be changed to 
complete = finishedTargetsCompleted.size() == targets.size()
return !complete;

File: twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/SourceReceiver.java [hunk 4]
```diff
@@ -218,21 +227,34 @@ public boolean progress() {
...
-          if (!needsFurtherProgress) {
+          boolean sendSuccess = !sendSyncForward(target);
+          if (sendSuccess) {
             targetStates.put(target, ReceiverState.SYNCED);
             // at this point we call the sync event
             onSyncEvent(target, barriers.get(target));
             clearTarget(target);
           }
+
+          if (!sendSuccess) {
```
Comment (chathurawidanage) [inline twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/SourceReceiver.java]: else

File: twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/partition/DPartitionBatchFinalReceiver.java (hunk 3 above)
Comment (supunkamburugamuve) [reply twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/partition/DPartitionBatchFinalReceiver.java]: Added

File: twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/SourceReceiver.java (hunk 4 above)
Comment (supunkamburugamuve) [reply twister2/comms/src/java/edu/iu/dsc/tws/comms/dfw/io/SourceReceiver.java]: Added

**Pull request merged**